File: csharp/react/React.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Reactor
{
    public InputCell CreateInputCell(int value) => new InputCell(value);
    public ComputeCell CreateComputeCell(IEnumerable<Cell> producers, Func<int[], int> compute)
        => new ComputeCell(producers, compute);
}

public abstract class Cell
{
    private int _value;
    public int Value
    {
        get => _value;
        set
        {
            if (_value != value)
            {
                _value = value;
                Changed?.Invoke(this, _value);
            }
        }
    }
    public event EventHandler<int> Changed; 
}

public class InputCell : Cell
{
    public InputCell(int value) => Value = value;
}

public class ComputeCell : Cell
{
    private IEnumerable<Cell> Producers { get; init; }
    private Func<int[], int> Compute { get; init; }
    public ComputeCell(IEnumerable<Cell> producers, Func<int[], int> compute)
    {
        (Producers, Compute) = (producers, compute);
        ComputeNew();
        foreach (var producer in OnlyInputCells())
            producer.Changed += (o,e) => ComputeNew(); // object and new integer aren't used by the callback
    }
    private void ComputeNew() =>  Value = Compute(Producers.Select(x => x.Value).ToArray());
    private IEnumerable<Cell> OnlyInputCells() 
        => Producers.SelectMany(x => x is ComputeCell c ? c.OnlyInputCells() : [(InputCell)x]).Distinct();
}

File: fsharp/react/React.fs
module React

[<AbstractClass>]
type Cell() =
    member val changed = new Event<int>()
    [<CLIEvent>]
    member this.Changed = this.changed.Publish
    abstract member Value : int with get, set

type InputCell(value: int) = 
    inherit Cell()
    let mutable number = value
    override this.Value 
        with get() = number
        and set(newValue: int) =
            if newValue <> number then 
                number <- newValue 
                this.changed.Trigger(number)

type ComputeCell(producers: Cell list, compute: int list -> int) as this =
    inherit Cell()
    let mutable number = 0
    do this.Recompute
    do this.InputProducers |> List.map (fun cell -> cell.Changed.Add (fun _ -> this.Recompute)) |> ignore
    override this.Value 
        with get() = number
        and set(newValue: int) = failwith "value cannot be set on a ComputeCell!"
    member this.Recompute = 
        let newValue = producers |> List.map (fun i -> i.Value) |> compute  
        if newValue <> number then 
            number <- newValue
            this.changed.Trigger(number)
    member this.InputProducers : InputCell list = 
        let isInputCell c : InputCell option = tryUnbox (box c)
        let getInputList c : InputCell list = 
            match isInputCell c with 
            | Some inputCell -> [inputCell]
            | None -> (unbox<ComputeCell> (box c)).InputProducers
        producers |> List.collect getInputList |> List.distinct

type Reactor() = 
    member _.createInputCell (value: int) = new InputCell(value)
    member _.createComputeCell (producers: Cell list) (compute: int list -> int) =
        new ComputeCell(producers, compute)